// @haxtheweb/clean-portfolio-theme/clean-portfolio-theme.js
/**
 * Copyright 2025 haxtheweb
 * @license Apache-2.0, see LICENSE for full text.
 */
import{LitElement as t,html as e,css as i}from"../../lit/index.js";import{HAXCMSLitElementTheme as o}from"../haxcms-elements/lib/core/HAXCMSLitElementTheme.js";import{store as r}from"../haxcms-elements/lib/core/haxcms-site-store.js";import{autorun as a,toJS as s}from"../../mobx/dist/mobx.esm.js";import"../haxcms-elements/lib/ui-components/active-item/site-active-title.js";import"../haxcms-elements/lib/ui-components/active-item/site-active-media-banner.js";import"../scroll-button/scroll-button.js";import"../simple-icon/lib/simple-icon-button-lite.js";import{DDDSuper as l}from"../d-d-d/d-d-d.js";import{DDDVariables as n}from"../d-d-d/lib/DDDStyles.js";import{licenseList as c}from"../license-element/license-element.js";const h=["https://fonts.googleapis.com/css2?family=Playfair+Display:ital,wght@0,400..900;1,400..900&family=Source+Code+Pro:ital,wght@0,200..900;1,200..900&family=Work+Sans:ital,wght@0,100..900;1,100..900&display=swap"];export class CleanPortfolioTheme extends(l(o)){static get tag(){return"clean-portfolio-theme"}constructor(){super(),this.siteTheme="",this.dataPrimary=2,this.activeLayout="text",this.activeParent="",this.selectedTag="",this.menuOpen=!1,this.menuOverflow=[],this.lastUpdated="",this.copyrightYear=0,this.categoryTags=[],this.allTags=[],this.items=[],this.__disposer=this.__disposer||[],a((t=>{this.homeLink=s(r.homeLink),this.__disposer.push(t)})),a((t=>{this.siteTitle=s(r.siteTitle),this.__disposer.push(t)})),a((t=>{this.ancestorItem=s(r.ancestorItem),this.__disposer.push(t)})),a((t=>{let e=s(r.activeTags);e&&e.length>0?this.activeTags=e.split(","):this.activeTags=[],this.__disposer.push(t)})),a((t=>{let e=r.getItemChildren(null);e&&e.length>0&&(this.topItems=[...e]),this.__disposer.push(t)})),a((t=>{const e=s(r.activeItem);if(e){let t=r.manifest.items.find((t=>e.parent===t.id));t||(t=""),globalThis.document&&globalThis.document.startViewTransition?globalThis.document.startViewTransition((()=>{this.activeItem=e,this.activeParent=t})):(this.activeItem=e,this.activeParent=t);const i=r.getItemChildren(r.activeId);if(i)if(i.length>0){this.setLayout("listing");const t=[],e=[];i.forEach((i=>{let o=s(i.metadata.tags);if(o){const i=o.split(",");i[0]&&!t.includes(i[0])&&t.push(i[0]),i.forEach((t=>{t&&!e.includes(t)&&e.push(t)}))}})),globalThis.document&&globalThis.document.startViewTransition?globalThis.document.startViewTransition((()=>{this.items=[...i],this.categoryTags=[...t],this.allTags=[...e]})):(this.items=[...i],this.categoryTags=[...t],this.allTags=[...e]),this.selectedTag="";const o=this.shadowRoot?.querySelector("#listing-select");o&&(o.value="")}else e.parent?this.setLayout("media"):this.setLayout("text")}this.__disposer.push(t)})),a((t=>{if(s(r.editMode)){const t=this.shadowRoot.querySelector("#slot")||globalThis.document.querySelector("#slot");t?.scrollIntoView({behavior:"smooth",block:"start"})}this.__disposer.push(t)})),a((t=>{this.manifest=s(r.manifest);let e=new c;this.manifest.license&&e[this.manifest.license]&&(this.licenseName=e[this.manifest.license].name,this.licenseLink=e[this.manifest.license].link,this.licenseImage=e[this.manifest.license].image),this.__disposer.push(t)}))}firstUpdated(t){super.firstUpdated(t);let e=globalThis.DesignSystemManager.requestAvailability();e.addDesignSystem({name:"clean-portfolio-theme",styles:[...CleanPortfolioTheme.styles,n],fonts:h,hax:!0}),e.active="clean-portfolio-theme",this.lastUpdated=new Date(1e3*r.manifest.metadata.site.updated).toDateString(),this.copyrightYear=new Date(1e3*r.manifest.metadata.site.created).getFullYear();const i=this.renderRoot.querySelector("nav");i&&(this._resizeObserver=new ResizeObserver((()=>{this._checkOverflow()})),this._resizeObserver.observe(i)),requestAnimationFrame((()=>this._checkOverflow()))}disconnectedCallback(){if(this._resizeObserver&&this._resizeObserver.disconnect(),this.__disposer)for(var t in this.__disposer)this.__disposer[t].dispose();super.disconnectedCallback()}_checkOverflow(){const t=this.renderRoot.querySelector("nav");if(t){const e=Array.from(t.children),i=t.clientWidth-100;let o=0;const r=[];for(const t of e)t.style.display="inline-block",o+=t.offsetWidth+25,o>i&&(t.style.display="none",r.push(this.topItems.find((e=>e.slug===t.getAttribute("href")))));this.menuOverflow=r,this.requestUpdate()}}static get properties(){return{...super.properties,activeItem:{type:Object},activeParent:{type:Object},ancestorItem:{type:Object},topItems:{type:Array},items:{type:Array},activeTags:{type:Array},categoryTags:{type:Array},allTags:{type:Array},menuOverflow:{type:Array},menuOpen:{type:Boolean},copyrightYear:{type:Number},siteTitle:{type:String},homeLink:{type:String},activeLayout:{type:String},selectedTag:{type:String},lastUpdated:{type:String},licenseName:{type:String},licenseLink:{type:String},licenseImage:{type:String},dataPrimary:{type:String,attribute:"data-primary",reflect:!0},siteTheme:{type:String,reflect:!0,attribute:"site-theme"}}}HAXCMSGlobalStyleSheetContent(){return[...super.HAXCMSGlobalStyleSheetContent(),i`
        :root, html, body {
            /* site color variables */
            --portfolio-black: #000000;
            --portfolio-white: #ffffff;
            --portfolio-lighterGrey: #bfbfbf;
            --portfolio-lightGrey: #8a8a8a;
            --portfolio-grey: #424242;
            --portfolio-darkGrey: #333;
            --portfolio-textHeader: #efefef;
            --portfolio-textHeaderHover: #b7b7b7;
            --portfolio-backgroundWhite: #f7f7f7;
            --portfolio-menuItemUnderline: #ff0000;
            --portfolio-cardTag: #6D4C41;
            --portfolio-cardTagLight: #c2a399;

            --portfolio-lightDark-blackWhite: light-dark(var(--portfolio-black), var(--portfolio-white));
            --portfolio-lightDark-whiteBlack: light-dark(var(--portfolio-white), var(--portfolio-black));
            --portfolio-lightDark-bg: light-dark(var(--portfolio-backgroundWhite), var(--portfolio-grey));
            --portfolio-lightDark-footer: light-dark(var(--portfolio-black), var(--portfolio-darkGrey));
            --portfolio-lightDark-cardTag: light-dark(var(--portfolio-cardTag), var(--portfolio-cardTagLight));
            --portfolio-lightDark-cardImg: light-dark(var(--portfolio-lighterGrey), var(--portfolio-darkGrey));

            --portfolio-breadcrumb: var(--portfolio-lightDark-blackWhite);

            /* site font variables */
            --portfolio-font-body: "Source Code Pro", system-ui, Monaco, Consolas, "Lucida Console", monospace;
            --portfolio-font-header: "Work Sans", system-ui, -apple-system, BlinkMacSystemFont, "Roboto", "Segoe UI", "Helvetica Neue", "Lucida Grande", Arial, sans-serif;

            --portfolio-fontsize-responsive: clamp(16px, 2vw, 22px);

            background-color: var(--portfolio-lightDark-bg);
            font-family: var(--portfolio-font-body);
            font-size: var(--portfolio-fontsize-responsive);
            color-scheme: light dark;
            scroll-behavior: smooth;
            /* theme colors */
            --portfolio-earth-accentLight: #689f38;
            --portfolio-earth-accentDark: #33691e;
            --portfolio-water-accentLight: #2a95cf;
            --portfolio-water-accentDark: #1e53a2;
            --portfolio-fire-accentLight: #ef5350;
            --portfolio-fire-accentDark: #8e2424;
            --portfolio-sand-accentLight: #f57c00;
            --portfolio-sand-accentDark: #6d4c41;
            --portfolio-rose-accentLight: #ef50a2;
            --portfolio-rose-accentDark:  #6a1b4d;
            --portfolio-violet-accentLight: #a27dff;
            --portfolio-violet-accentDark:  #392b6a;
        }

        body.dark-mode {
          color-scheme: only dark;
        }
        `]}static get styles(){return[super.styles,i`
      /* Semantic elements */
      :host {
        display: block;
      }

      header {
        background-color: var(--portfolio-darkGrey);
        transition: all 0.2s ease-in-out;
        padding: 22px;
      }

      .header-inner {
        display: flex;
        align-items: center;
        width: 100%;
        max-width: 1236px;
        margin: 0 auto;
        line-height: 1.7;
        position: relative;
      }

      nav {
        display: flex;
        gap: .5rem;
        justify-content: flex-end;
        flex: 1 1 auto;
        min-width: 0;
      }

      h1, h2, h3, h4, h5, h6 {
        color: var(--portfolio-lightDark-blackWhite);
        line-height: 1.2;
        font-weight: 700;
        margin-bottom: 0.75rem;
        margin-top: 0;
      }
      
      p {
        font-size: var(--portfolio-fontsize-responsive);
        line-height: 1.5;
        margin-bottom: 1.3em;
      }

      ul {
        list-style-type: disc;
      }

      li {
        font-family: var(--portfolio-font-body);
      }

      li::marker {
        color: var(--portfolio-lightDark-blackWhite) !important;
      }

      footer {
        display: flex;
        justify-content: space-evenly;
        align-items: center;
        background-color: var(--portfolio-darkGrey);
        color: var(--portfolio-white);
        font-family: var(--portfolio-font-body);
        font-size: clamp(11px, 2vw, 15px);
        padding: 40px 5vw;
        margin-top: 150px;
        transition: .3s;
      }

      :focus {
        outline-offset: 8px;
      }

      /* Site header elements */
      #site-title {
        padding: 0.5rem;
        color: var(--portfolio-textHeader);
        font-family: var(--portfolio-font-header);
        font-weight: bold;
        font-size: var(--portfolio-fontsize-responsive);
        text-transform: uppercase;
        text-decoration: none;
        border: 5px solid white;
        transition: all 0.2s ease-in-out;
      }

      #site-title:hover,
      #site-title:focus {
        color: var(--portfolio-textHeaderHover);
      }

      header a.menu-item {
        display: inline-block;
        position: relative;
        margin: 10px;
        padding: 0 5px;
        white-space: nowrap;
        color: var(--portfolio-textHeader);
        font-family: var(--portfolio-font-header);
        font-size: var(--portfolio-fontsize-responsive);
        font-weight: 450;
        text-align: center;
        text-decoration: none;
        transition: all 0.3s ease-in-out;
      }

      header a.menu-item:after {
        content: "";
        position: absolute;
        bottom: -10px;
        left: 50%;
        width: 0;
        height: 4px;
        background-color: var(--portfolio-menuItemUnderline);
        transform: translateX(-50%);
        transition: all 0.2s ease-in-out;
      }

      .menu-item:hover,
      .menu-item:focus,
      .menu-item.active {
        color: var(--portfolio-textHeaderHover);
      }

      .menu-item:hover:after,
      .menu-item:focus:after,
      .menu-item.active:after {
        width: 100%;
        transform: translateX(-50%) scaleX(1);
      }

      /* Mobile menu */
      .mobile-menu-wrapper {
        position: relative;
        display: flex;
        justify-content: center;
      }

      header button {
          display: inline-block;
          color: #fff;
          background-color: var(--portfolio-lightGrey);
          border: 0;
          width: 45px;
          height: 50px;
          cursor: pointer;
          margin-left: 10px;
          transition: all 0.2s ease-in-out;
      }

      .navicon {
          position: relative;
          width: 1.5rem;
          height: .25rem;
          background: #fff;
          margin: auto;
          transition: .3s;
      }

      .navicon:before,
      .navicon:after {
          content: "";
          position: absolute;
          left: 0;
          width: 1.5rem;
          height: .25rem;
          background: #fff;
          transition: .3s;
      }

      .navicon:before {
          top: -0.5rem;
      }

      .navicon:after {
          bottom: -0.5rem;
      }

      .close .navicon {
          background: rgba(0,0,0,0);
      }

      .close .navicon:before,
      .close .navicon:after {
          transform-origin: 50% 50%;
          top: 0;
          width: 1.5rem;
      }

      .close .navicon:before {
          transform: rotate3d(0, 0, 1, 45deg);
      }

      .close .navicon:after {
          transform: rotate3d(0, 0, 1, -45deg);
      }

      .hidden-links {
        position: absolute;
        top: calc(100% + 12px);
        right: 0;
        padding: 5px;
        gap: 0;
        border-radius: 4px;
        box-shadow: 0 2px 4px 0 rgba(#000, 0.16), 0 2px 10px 0 rgba(#000, 0.12);
        background-color: var(--portfolio-lightGrey);
        z-index: 5;
        transition: .3s;
      }

      .hidden-links:before {
        content: "";
        position: absolute;
        top: -8px;
        right: 12px;
        border-style: solid;
        border-width: 0 10px 10px;
        border-color: var(--portfolio-lightGrey) transparent;
        transition: .3s;
      }

      .hidden-links li {
        display: block;
        border-bottom: 1px solid var(--portfolio-lighterGrey);
        padding: 10px;
      }

      .hidden-links li:last-child {
        border-bottom: none;
      }

      .hidden-links a {
        color: var(--portfolio-white);
        font-size: 16px;
        font-weight: 400;
        font-family: var(--portfolio-font-header);
        text-decoration: none;
        display: block;
        transition: .3s;
      }

      .hidden-links a:hover,
      .hidden-links a:focus,
      .hidden-links a.active {
        color: var(--portfolio-lighterGrey);
      }

      .hidden {
        display: none;
      }

      .visually-hidden,
      .screen-reader-text,
      .screen-reader-text span,
      .screen-reader-shortcut {
        position: absolute !important;
        clip: rect(1px, 1px, 1px, 1px);
        height: 1px !important;
        width: 1px !important;
        border: 0 !important;
        overflow: hidden;
      }

      /* Imported elements */
      site-active-title h1 {
        font-family: "Playfair Display";
        font-size: 88px;
        font-weight: bold;
        transition: .3s;
        view-transition-name: location;
      }

      site-active-media-banner {
        --media-banner-background-color: none;
        --media-banner-height: clamp(320px, 80vw, 640px);
        margin: 0;
      }

      .theme-picker {
        --simple-icon-width: 24px;
        --simple-icon-height: 24px;
        padding: 0;
        border: solid 3px var(--portfolio-lightGrey);
        border-radius: 50%;
        background-color: var(--portfolio-white);
        z-index: 4;
        color: var(--portfolio-darkGrey);
        transition: color .3s, border .3s;
      }

      header .theme-picker {
        position: absolute;
        top: 8px;
        right: 6px;
      }

      .breadcrumb .theme-picker {
        visibility: hidden;
        margin-left: auto;
        margin-right: 8px;
      }

      /* Tags */
      .tag-list {
        flex-direction: row;
        flex-wrap: wrap;
        list-style-type: none;
        margin: 0 0 20px 0;
        padding: 0;
        border-top: 3px solid var(--portfolio-lightDark-blackWhite);
        border-bottom: 1px solid var(--portfolio-lightDark-blackWhite);
      }

      .tag-list li {
        margin: 0 10px 0 0;
        padding-top: 5px;
        padding-bottom: 5px;
      }

      .tag-list li a {
        color: var(--portfolio-lightDark-blackWhite);
        font-weight: 400;
        font-size: 1.25em;
        text-transform: uppercase;
      }

      /* Breadcrumb */
      .breadcrumb-wrapper {
        max-width: 1236px;
        margin: 0 auto;
        padding: 0 22px;
        view-transition-name: location;
      }

      .breadcrumb {
        display: flex;
        align-items: center;
        gap: 10px;
        height: clamp(40px, 5vw, 60px);
        color: var(--portfolio-lightDark-blackWhite);
        font-family: var(--portfolio-font-header);
        font-size: clamp(14px, 2vw, 18px);
      }

      .breadcrumb a {
        color: var(--portfolio-breadcrumb);
        text-decoration-color: var(--portfolio-breadcrumb);
        transition: 0.3s;
        font-weight: 450;
      }

      .breadcrumb-parent {
        color: var(--portfolio-lightDark-blackWhite);
      }

      .breadcrumb-split {
        color: var(--portfolio-breadcrumb);
        transition: .3s;
      }

      .breadcrumb-title {
        font-weight: bold;
      }
      
      /* Layouts */
      .container {
        margin: 64px auto;
        width: 90%;
        max-width: 840px;
        text-align: left;
        view-transition-name: location;
      }

      #contentcontainer {
        min-height: 50vh;
      } 

      #slot {
        min-height: 0vh;
      }

      .listing-titlecontainer {
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-wrap: wrap;
        gap: 12px;
        margin-bottom: 20px;
        view-transition-name: location;
      }

      #listing-select {
        font-size: 1rem;
        padding: 0.5em;
        max-width: 240px;
        width: 100%;
      }

      .listing-category {
        font-family: var(--portfolio-font-body);
        font-size: var(--portfolio-fontsize-responsive);
        text-transform: uppercase;
        border-top: 6px solid var(--portfolio-lightDark-blackWhite);
        min-height: 36px;
      }

      .listing-category a {
        color: var(--portfolio-lightDark-blackWhite);
      }

      .listing-grid {
        display: grid;
        gap: 24px;
        width: 100%;
        view-transition-name: location;
        grid-template-columns: repeat(4, 1fr);
      }

      div .listing-card {
        width: 188px;
        margin-bottom: 48px;
        text-decoration: none;
        transition: .4s;
      }

      .listing-cardimg {
        background-color: var(--portfolio-lightDark-cardImg);
        border-radius: 6px;
        margin-bottom: 12px;
        height: 120px;
        transition: all 0.2s ease-in-out;
        overflow: hidden;
        display: block;
      }

      .listing-card:hover .listing-cardtitle,
      .listing-card:focus .listing-cardtitle{
        text-decoration: underline;
      }

      .listing-card:hover .listing-cardimg,
      .listing-card:focus .listing-cardimg {
        box-shadow: rgba(0, 0, 0, 0.25) 0px 0px 10px 0px;
      }

      .listing-cardimg img {
        width: 100%;
        height: 100%;
        object-fit: cover;
      }

      .listing-cardtitle {
        color: var(--portfolio-lightDark-blackWhite);
        font-family: var(--portfolio-font-header);
        font-weight: 400;
        font-size: clamp(15px, 2vw, 21px);
        text-transform: uppercase;
        margin-bottom: 5.5px;
      }

      .listing-cardtag {
        color: var(--portfolio-lightDark-cardTag);
        font-family: var(--portfolio-font-body);
        font-size: clamp(8px, 2vw, 14px);
        font-weight: 400;
        transition: .3s;
      }

      /* Media queries */
      @media (max-width: 60em) {
        .listing-grid {
          grid-template-columns: repeat(2, 1fr);
        }
        div .listing-card {
          width: 100%;
          margin-bottom: 36px;
        }
        .listing-cardimg {
          height: 210px;
        }
        site-active-title h1 {
          font-size: 72px;
        }
        header .theme-picker {
          visibility: hidden;
        }
        .breadcrumb .theme-picker {
          visibility: visible;
        }
      }

      @media (max-width: 37.5em) {
        .listing-grid {
          grid-template-columns: repeat(1, 1fr);
        }
        div .listing-card {
          margin-bottom: 24px;
        }
        .listing-cardimg {
          height: 240px;
        }
        site-active-title h1 {
          font-size: clamp(56px, 12vw, 64px);
        }
      }

      /* Theme colors */
      :host([site-theme="earth"]) {
        --portfolio-menuItemUnderline:var(--portfolio-earth-accentLight);
        --portfolio-lightGrey: var(--portfolio-earth-accentLight);
        --portfolio-breadcrumb: var(--portfolio-earth-accentLight);
        --portfolio-darkGrey: var(--portfolio-earth-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-earth-accentDark),
          var(--portfolio-earth-accentLight)
        );
      }

      :host([site-theme="water"]) {
        --portfolio-menuItemUnderline:var(--portfolio-water-accentLight);
        --portfolio-lightGrey: var(--portfolio-water-accentLight);
        --portfolio-breadcrumb: var(--portfolio-water-accentLight);
        --portfolio-darkGrey: var(--portfolio-water-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-water-accentDark),
          var(--portfolio-water-accentLight)
        );
      }

      :host([site-theme="fire"]) {
        --portfolio-menuItemUnderline:var(--portfolio-fire-accentLight);
        --portfolio-lightGrey: var(--portfolio-fire-accentLight);
        --portfolio-breadcrumb: var(--portfolio-fire-accentLight);
        --portfolio-darkGrey: var(--portfolio-fire-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-fire-accentDark),
          var(--portfolio-fire-accentLight)
        );
      }

      :host([site-theme="sand"]) {
        --portfolio-menuItemUnderline:var(--portfolio-sand-accentLight);
        --portfolio-lightGrey: var(--portfolio-sand-accentLight);
        --portfolio-breadcrumb: var(--portfolio-sand-accentLight);
        --portfolio-darkGrey: var(--portfolio-sand-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-sand-accentDark),
          var(--portfolio-sand-accentLight)
        );
      }

      :host([site-theme="rose"]) {
        --portfolio-menuItemUnderline:var(--portfolio-rose-accentLight);
        --portfolio-lightGrey: var(--portfolio-rose-accentLight);
        --portfolio-breadcrumb: var(--portfolio-rose-accentLight);
        --portfolio-darkGrey: var(--portfolio-rose-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-rose-accentDark),
          var(--portfolio-rose-accentLight)
        );
      }

      :host([site-theme="violet"]) {
        --portfolio-menuItemUnderline:var(--portfolio-violet-accentLight);
        --portfolio-lightGrey: var(--portfolio-violet-accentLight);
        --portfolio-breadcrumb: var(--portfolio-violet-accentLight);
        --portfolio-darkGrey: var(--portfolio-violet-accentDark);
        --portfolio-lightDark-cardTag: light-dark(
          var(--portfolio-violet-accentDark),
          var(--portfolio-violet-accentLight)
        );
      }
    `]}testEditMode(t){this.editMode&&(t.preventDefault(),t.stopPropagation(),t.stopImmediatePropagation())}_renderListing(){const t=this.selectedTag?this.items.filter((t=>{let e=s(t.metadata.tags);return!!e&&e.includes(this.selectedTag)})):this.items,i=this.selectedTag?this.categoryTags.filter((e=>t.some((t=>{let i=s(t.metadata.tags);return!!i&&i.includes(e)})))):this.categoryTags;return e`
      <!-- Render cards based on filtered tags -->
      ${i.length>0?i.map((i=>e`
              <h3 class="listing-category"><a @click="${this.testEditMode}" href="x/tags?display=card&tag=${i.trim()}">${i}</a></h3>
              <div class="listing-grid">
                ${t.filter((t=>{let e=s(t.metadata.tags);return!!e&&e.includes(i)})).map((t=>{let i=s(t.metadata.tags).split(",")[1];return e`
                    <a class="listing-card" href="${t.slug}">
                      <div class="listing-cardimg">
                        <img src="${t.metadata.image}" onerror="this.style.display='none'">
                      </div>
                      <div class="listing-cardtitle">${t.title}</div>
                      <div class="listing-cardtag">${i}</div>
                    </a>
                  `}))}
              </div>
            `)):e``}

      <!-- Render cards with no tags -->
      ${!this.selectedTag&&this.items.some((t=>!s(t.metadata.tags)))?e`
        <h3 class="listing-category"></h3>
        <div class="listing-grid">
          ${this.items.filter((t=>!s(t.metadata.tags))).map((t=>e`
            <a class="listing-card" href="${t.slug}">
              <div class="listing-cardimg">
                <img src="${t.metadata.image}" onerror="this.style.display='none'">
              </div>
              <div class="listing-cardtitle">${t.title}</div>
            </a>
          `))}
        </div>
      `:""}
    `}setLayout(t){globalThis.document&&globalThis.document.startViewTransition?globalThis.document.startViewTransition((()=>{this.activeLayout=t})):this.activeLayout=t}toggleSiteTheme(t){switch(t.currentTarget.blur(),this.siteTheme){case"earth":this.siteTheme="water",this.dataPrimary=1;break;case"water":this.siteTheme="fire",this.dataPrimary=11;break;case"fire":this.siteTheme="sand",this.dataPrimary=23;break;case"sand":this.siteTheme="rose",this.dataPrimary=35;break;case"rose":this.siteTheme="violet",this.dataPrimary=47;break;case"violet":this.siteTheme="",this.dataPrimary=2;break;default:this.siteTheme="earth",this.dataPrimary=1}}render(){return e`
      <header>
        <div class="header-inner">
          <a id="site-title" @click=${t=>t.currentTarget.blur()} href="${this.homeLink}">${this.siteTitle}</a>
          <nav>
            ${this.topItems.map((t=>e`
                  <a
                    class="menu-item ${this.activeItem&&(t.id===this.activeItem.id||this.ancestorItem&&t.id===this.ancestorItem.id)?"active":""}"
                    href="${t.slug}"
                  >
                    ${t.title}
                  </a>
                `))}
          </nav>
          ${this.menuOverflow.length>0?e`
              <div class="mobile-menu-wrapper">
                <button type="button" class=${this.menuOpen?"close":""} @click=${()=>this.menuOpen=!this.menuOpen}>
                  <span class="visually-hidden">Toggle Menu</span>
                  <div class="navicon"></div>
                </button>
                <ul class="hidden-links ${this.menuOpen?"":"hidden"}">
                  ${this.menuOverflow.map((t=>e`
                    <li>
                      <a
                        class="${this.activeItem&&(t.id===this.activeItem.id||this.ancestorItem&&t.id===this.ancestorItem.id)?"active":""}"
                        href="${t.slug}"
                        @click=${t=>t.currentTarget.blur()}
                      >
                        ${t.title}
                      </a>
                    </li>`))}
                </ul>
              </div>
            `:""}
        </div>
        <simple-icon-button-lite icon="image:style" class="theme-picker" @click="${this.toggleSiteTheme}"></simple-icon-button-lite>
      </header>
      
      <div class="breadcrumb-wrapper">
        <div class="breadcrumb">
          ${this.activeParent?e`
              <a href=${this.activeParent.slug}>
                <span>←</span>
                <span class="breadcrumb-parent">${this.activeParent.title}</span>
              </a>
              <span class="breadcrumb-split">/</span>
              <span class="breadcrumb-title">${this.activeItem.title}</span>
            `:""}
          <simple-icon-button-lite icon="image:style" class="theme-picker" @click="${this.toggleSiteTheme}"></simple-icon-button-lite>
        </div>
      </div>

      ${"media"==this.activeLayout?e`<site-active-media-banner></site-active-media-banner>`:""}

      <div id="contentcontainer" class="container">
        ${"listing"==this.activeLayout?e`
              <div class="listing-titlecontainer">
                <site-active-title></site-active-title>
                <!-- Render select for filtering tags (only appears if >1 tag OR 1 tag and items with no tag) -->
                ${this.categoryTags.length>1||this.categoryTags.length>0&&this.items.some((t=>!t.metadata.tags))?e`
                    <select id="listing-select" @change=${t=>this.selectedTag=t.target.value}>
                      <option value="" ?selected="${""===this.selectedTag}">All</option>
                      ${this.categoryTags.map((t=>e`
                          <option value="${t}" ?selected="${this.selectedTag===t}">
                            ${t}
                          </option>`))}
                    </select>`:""}
              </div>
          `:e`
                <site-active-title></site-active-title>
                ${this.activeTags&&this.activeTags.length>0?e`
                      <ul class="tag-list">
                        ${("listing"===this.activeLayout?this.activeTags:this.activeTags.slice(1)).map((t=>e`<li><a @click="${this.testEditMode}" href="x/tags?display=card&tag=${t.trim()}">${t}</a></li>`))}
                      </ul>
                  `:""}
          `}

        ${"listing"==this.activeLayout?this._renderListing():""}

        <div id="slot"><slot></slot></div>
      </div>

      <footer>
        Generated: ${this.lastUpdated}<br><br>
        © ${this.copyrightYear} ${r.manifest.author}.
        <div
          class="license-body"
          xmlns:cc="${this.licenseLink}"
          xmlns:dc="http://purl.org/dc/elements/1.1/"
        >
          ${this.licenseImage?e`
                <a
                  class="big-license-link"
                  target="_blank"
                  href="${this.licenseLink}"
                  rel="noopener noreferrer"
                >
                  <img
                    loading="lazy"
                    decoding="async"
                    fetchpriority="low"
                    alt="${this.licenseName} graphic"
                    src="${this.licenseImage}"
                  />
                </a>
              `:""}
        </div>
        <simple-icon-button-lite icon="image:style" class="theme-picker" @click="${this.toggleSiteTheme}"></simple-icon-button-lite>
        <scroll-button @click=${t=>t.currentTarget.blur()}></scroll-button>
      </footer>
    `}}globalThis.customElements.define(CleanPortfolioTheme.tag,CleanPortfolioTheme);
